clang/test/SemaSYCL/implicit-float-size-conversion.cpp

```diff
@@ -0,0 +1,50 @@
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64 -internal-isystem %S/Inputs -fsyntax-only -sycl-std=2020 -Wimplicit-float-size-conversion -verify -DCHECK_SIZE %s
```

Comment:
Same suggestion here as the one @elizabethandrews made above.

Reply:
Custom prefixes like `-verify=size,precision` , `-verify=size, -verify=precision` etc doesn't seem to support testing of multiple RUN commands each with a different set of warning flags.

Example:  
```
// RUN: ... -Wimplicit-float-size-conversion -verify=size %s
// RUN: ... -Wimplicit-float-conversion -verify=size,precision %s

q.submit([&](handler &h) {
    // size-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<kernelA, (lambda}}
    h.single_task<class kernelA>([=]() {
      float s = 1.0; // size-warning {{implicit conversion between floating point types of different sizes}}
      float d = 2.1; // size-warning {{implicit conversion between floating point types of different sizes}}
    });
  });

  q.submit([&](handler &h) {
    // precision-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<kernelA, (lambda}}
    h.single_task<class kernelA>([=]() {
      float s = 1.0; // size-warning {{implicit conversion between floating point types of different sizes}}
      float d = 2.1; // precision-warning {{implicit conversion loses floating-point precision: 'double' to 'float'}} // We get 
      an additional size warning here
    });
  });

```

Reply:
I'm not sure I understand the problem you're running into. You should be able to:
```
      float d = 2.1; // precision-warning {{implicit conversion loses floating-point precision: 'double' to 'float'}} \
                        size-warning {{implicit conversion between floating point types of different sizes}}
```

Reply:
In the original test, we were testing individual cases of passing `-Wimplicit-float-size-conversion`, `-Wimplicit-float-conversion` and a combination of both these flags.
The `RUN` commands were conditionally run on each set of test cases.
Now when I use custom prefixes, the RUN command gets applied to all of the code rather than specific blocks.
Previously  the following RUN command was used to test the behavior of `-Wimplicit-float-conversion` by itself and only the `-Wimplicit-float-conversion` flag got enabled.

```
// RUN: %clang_cc1 -fsycl-is-device -triple spir64 -internal-isystem %S/Inputs -fsyntax-only -sycl-std=2020 
-Wimplicit-float-conversion -verify -DCHECK_PRECISION %s

#elif defined(CHECK_PRECISION)
  q.submit([&](handler &h) {
    // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<kernelA, (lambda}}
    h.single_task<class kernelA>([=]() {
      float s = 1.0; // expected-warning {{implicit conversion between floating point types of different sizes}}
      float d = 2.1; // expected-warning {{implicit conversion loses floating-point precision: 'double' to 'float'}}
    });
  });
```

But with the custom prefixes, if the first RUN command looked like below, and we only want to test the behavior of  -`Wimplicit-float-conversion`, we still get the `-Wimplicit-float-size-conversion` flag enabled from the first RUN command
```
// RUN: %clang_cc1 -fsycl-is-device -triple spir64 -internal-isystem %S/Inputs -fsyntax-only -sycl-std=2020 -Wimplicit-float-size-conversion -verify=size %s
// RUN: %clang_cc1 -fsycl-is-device -triple spir64 -internal-isystem %S/Inputs -fsyntax-only -sycl-std=2020 -Wimplicit-float-conversion -verify=size,precision %s

q.submit([&](handler &h) {
    // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<kernelA, (lambda}}
    h.single_task<class kernelA>([=]() {
      float s = 1.0; // expected-warning {{implicit conversion between floating point types of different sizes}}
      float d = 2.1; // expected-warning {{implicit conversion loses floating-point precision: 'double' to 'float'}} // get the size warning as well because -Wimplicit-float-size-conversion gets applied from the first RUN
```

Reply:
> Now when I use custom prefixes, the RUN command gets applied to all of the code rather than specific blocks.

Correct -- we don't want conditional code in the test because that makes the test harder to read or modify.

> But with the custom prefixes, if the first RUN command looked like below, and we only want to test the behavior of -Wimplicit-float-conversion, we still get the -Wimplicit-float-size-conversion flag enabled from the first RUN command

If I'm understanding you properly, that has nothing to do with using `-verify=` in the RUN line. The issue is that `-Wimplicit-float-conversion` will enable `-Wimplicit-float-size-conversion` automatically because of the diagnostic grouping in: https://github.com/intel/llvm/pull/6323/files#diff-c21da66b943620c063559b855971be3b2622d5ba1e99e9ff6d9cf0bf132b9fceR115

If you don't want the size conversions enabled for the second line, you should use `-Wno-implicit-float-size-conversion` after enabling `-Wimplicit-float-conversion`. It's a good idea to have such a test! But you'll also want a RUN line that enables `-Wimplicit-float-conversion` by itself without disabling the size conversion warning, so that we can test the behavior of that situation to ensure it doesn't issue duplicate diagnostics, etc.

Reply:
>The issue is that -Wimplicit-float-conversion will enable `-Wimplicit-float-size-conversion` automatically because of the diagnostic grouping
Right, I am aware of that grouping. But when there is a precision change (`float d = 2.1;`), we wanted the precision warning to take prioty over the size warning and just emit the precision warning.  Did I get this right? There were never both precision and size warning when `-Wimplicit-float-conversion` was only enabled.
So for `float d = 2.1;` with just `-Wimplicit-float-conversion` enabled, we would expect only the precision warning. Atleast, this is what the original test was doing. [only -Wimplicit-float-conversion enabled test case](https://github.com/intel/llvm/blob/23e73be61c169480741fa778fe40d3dcee51464a/clang/test/SemaSYCL/implicit-float-size-conversion.cpp#L27)

Reply:
> But when there is a precision change (float d = 2.1;), we wanted the precision warning to take prioty over the size warning and just emit the precision warning. Did I get this right?

Yes, you have that right.

> So for float d = 2.1; with just -Wimplicit-float-conversion enabled, we would expect only the precision warning.

Agreed that's the behavior we want.

I'm truly sorry if I'm just being dense with what the issue is. It might help if you could show the exact test case and what behavior you get when you run lit over that test.

One thing I notice from earlier in the thread:

> ```
> // RUN: %clang_cc1 -fsycl-is-device -triple spir64 -internal-isystem %S/Inputs -fsyntax-only -sycl-std=2020 -Wimplicit-float-size-conversion -verify=size %s
> // RUN: %clang_cc1 -fsycl-is-device -triple spir64 -internal-isystem %S/Inputs -fsyntax-only -sycl-std=2020 -Wimplicit-float-conversion -verify=size,precision %s
> 
> q.submit([&](handler &h) {
>     // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<kernelA, (lambda}}
>     h.single_task<class kernelA>([=]() {
>       float s = 1.0; // expected-warning {{implicit conversion between floating point types of different sizes}}
>       float d = 2.1; // expected-warning {{implicit conversion loses floating-point precision: 'double' to 'float'}} // get the size warning as well because -Wimplicit-float-size-conversion gets applied from the first RUN
> ```
This is still using `expected-warning` instead of `size-warning` and `precision-warning`. Be sure your real test file is using the correct prefix from the `-verify` lines when testing the expected diagnostics.

Reply:
Thanks for pointing out that we could use distinct custom prefixes for differentiating warnings from different RUN lines. I have fixed the test.